### PR TITLE
feat(licenses): SPDX expression parser with AST for SBOM/legal compliance (#351)

### PR DESCRIPTION
**Pull request by kotakanbe**
## Summary

Adds a new domain-layer SPDX license expression parser at `internal/domain/licenses/expression.go`. **v2 redesign** preserves operator semantics (AND/OR/WITH/+) as a true AST, not a flat operand list — required for SBOM (CycloneDX/SPDX) round-trip output and legal-compliance use cases. Critical-path follow-up to PR #345 (Maven POM fallback) and a hard blocker for #354 (ClearlyDefined.io integration), which sees compound expressions in ~30% of `licensed.declared` responses.

## Public API

```go
type ExpressionResult struct { Raw string; Root *ExprNode }

type ExprNode struct { License *ExprLicense; Compound *ExprCompound }

type ExprLicense struct {
    Raw           string
    Identifier    string             // canonical SPDX, "" if non-SPDX
    Normalization NormalizationResult
    OrLater       bool               // "+" suffix
    Exception     string             // verbatim WITH operand
}

type ExprCompound struct { Operator string; Operands []*ExprNode }

func ParseExpression(raw string) ExpressionResult
func (n *ExprNode) String() string                  // canonical SPDX round-trip
func (r ExpressionResult) Leaves() []*ExprLicense   // flat enumeration
```

## Grammar (SPDX 2.3 precedence: WITH > AND > OR)

```
expression = and_expr (OR and_expr)*
and_expr   = with_expr (AND with_expr)*
with_expr  = primary (WITH ident)?
primary    = ident ['+'] | '(' expression ')'
```

Same-operator chains flatten ("A OR B OR C" and "A OR (B OR C)" both produce one Compound with three children — set-equivalent in legal terms, matches SPDX renderer convention).

## Examples

| Input | AST shape | `String()` |
|---|---|---|
| `Apache-2.0` | leaf | `Apache-2.0` |
| `Apache License 2.0` | leaf (alias→canonical) | `Apache-2.0` |
| `Apache-2.0 OR MIT` | Compound{OR, [A, M]} | `Apache-2.0 OR MIT` |
| `Apache-2.0 OR MIT AND BSD-3-Clause` | Compound{OR, [A, Compound{AND, [M, B]}]} | `Apache-2.0 OR (MIT AND BSD-3-Clause)` |
| `(Apache-2.0 OR MIT) AND BSD-3-Clause` | Compound{AND, [Compound{OR, [A, M]}, B]} | `(Apache-2.0 OR MIT) AND BSD-3-Clause` |
| `Apache-2.0+ WITH Classpath-exception-2.0` | leaf{OrLater, Exception} | `Apache-2.0+ WITH Classpath-exception-2.0` |
| `(GPL-2.0-only OR Apache-2.0) WITH Classpath-exception-2.0` | distributes exception to each leaf | `GPL-2.0-only WITH Classpath-exception-2.0 OR Apache-2.0 WITH Classpath-exception-2.0` |
| `EPL-1.0 OR (LGPL-2.1 OR LGPL-3.0)` | Compound{OR, [3 leaves]} | `EPL-1.0 OR LGPL-2.1 OR LGPL-3.0` |

## Robustness

- Empty / whitespace-only / oversized (>64KB) input → `Root` is nil
- Stray edge operators (`OR Apache-2.0`, `Apache-2.0 OR`, `OR OR MIT`) silently dropped per v1 contract
- `Compound` with fewer than 2 children collapses to single child or nil — never emits degenerate compound
- `(A OR B) WITH X` — SPDX-strict-illegal but real in Maven / ClearlyDefined data — distributes the exception to every leaf so the legally-significant clause is not lost
- Unbalanced parens tolerated; parser never returns errors

## Programmer-error invariants

`ExprNode.String()` and `ExpressionResult.Leaves()` panic on:
- Zero-valued `ExprNode` (neither License nor Compound)
- `nil` operand inside a `Compound`

These shapes can only arise from hand-built ASTs; the parser itself never produces them. Surfacing them as panics catches misuse at the first render rather than silently producing malformed SPDX text.

## Idempotence

`ParseExpression(s).Root.String()` is a fixed point under re-parsing:
re-parsing the rendered string and rendering again produces the same text (verified by `TestExprNode_StringIdempotent`). This is the contract SBOM serialization callers depend on.

## Test plan

- [x] AST shape: leaf, mixed-operator precedence, paren grouping, flattening
- [x] Round-trip idempotence on canonical and edge-case inputs
- [x] NOASSERTION / alias / casefold / heuristic / exact MatchType coverage
- [x] WITH chain truncation pinned (`A WITH B WITH C` → first only)
- [x] `(Compound) WITH X` distributes exception; pre-existing inner exception preserved
- [x] `+` placement: trailing single → OrLater; bare / `++` → not recognized
- [x] Combined +/WITH preserves both flags through round-trip
- [x] Free-text license name with adjacent operator (`Apache License 2.0 OR MIT`) splits correctly
- [x] Hyphenated identifiers (`OR-tools`, `AND-license`) not split
- [x] Length cap boundary (input == cap accepted; cap+1 rejected)
- [x] Deep nesting (512 paren levels) and long flat chains (1024 operands) terminate
- [x] Panic invariants for zero-valued and nil-operand nodes (in both `String()` and `Leaves()`)
- [x] Asymmetric unbalanced-paren handling pinned
- [x] `go vet`, `golangci-lint` clean; `go test ./...` clean under `-race`

Refs #327, refs #351 (this PR lands the parser primitive only; the per-operand `ResolvedLicense` emission and Maven POM dispatcher first-SPDX promotion that #351 ultimately requires are deferred to a follow-up wiring PR).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by kotakanbe** [comment]
@claude Follow the procedure in `.github/prompts/review.prompt.md` with `--copilot-only` mode for this PR.

Phase 2: Resolve all unresolved Copilot review threads (discover → classify → fix → build/test/lint → commit/push → reply/resolve).
Phase 3: Detect recurring patterns from FIX-classified comments, propose and apply coding rules to `.github/instructions/`, then run `make sync-instructions`.
Phase 4: Post a summary comment on this PR.

**Comment by kotakanbe** [comment]
**Claude finished @kotakanbe's task in 3m 45s** —— [View job](https://github.com/future-architect/uzomuzo-oss/actions/runs/25133120444)

---
### Copilot Review Resolution (`--copilot-only`)

- [x] Gather context — read review prompt and affected files
- [x] Phase 2 — Discover, classify, and resolve Copilot comments
- [x] Phase 3 — Skipped (merge conflicts with main in source code)
- [x] Phase 4 — Summary

---

### Copilot Resolution — PR #358

| # | File | Classification | Action |
|---|------|---------------|--------|
| 1 | `expression.go:70` | **FIX** | Corrected v1 limitation doc: `+` suffix resolves to SPDX via alias table, not non-SPDX |

Commit: `8a5e597`
Threads resolved: 1/1

### Rules Learned

| Pattern | Status | Detail |
|---------|--------|--------|
| comment-doc-drift | **Skipped** | Phase 3 skipped: PR branch has merge conflicts with main in source code files. Rebase manually before re-running `/review --copilot-only` to record the pattern. |

**Note:** The FIX-classified comment falls under the `comment-doc-drift` category (doc comment claims behavior that contradicts the actual code and test suite). This is already partially covered by the existing **Comment-Code Consistency** rule in `copilot-learned-coding.instructions.md`, so promotion would likely be a refinement rather than a new rule.

---

File: internal/domain/licenses/expression.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+// Limitations (v1):
+//   - Mixed AND/OR is flattened into a single operand list. Operator structure
+//     and precedence are not preserved (no AST). Consumers that need to
+//     distinguish AND vs OR semantics for legal compliance must not rely on
+//     this parser alone.
+//   - "WITH" clauses stay attached to the preceding license-id; the combined
+//     "license-id WITH exception-id" string is normalized as a unit and will
+//     typically be reported as non-SPDX since no exception table exists yet.
+//   - The "+" suffix on a license-id (e.g., "Apache-2.0+") is preserved in Raw;
+//     normalization falls through to the heuristic, so the result is non-SPDX.
```
[the remark this answers is not in the document]
Comment (kotakanbe) [reply internal/domain/licenses/expression.go]: Fixed in 8a5e597.
